### ISSUE TITLE
Argo UI does not show complete logs 

### ISSUE THREAD

**Issue by Atharex** [argoproj/argo-workflows#5146]
## Summary 

Getting similar problems like in #3065
The logs are wholly seen in Minio saved logs and in the pod logs with kubectl, but not in the Argo UI. Neither of them show any errors, only the Argo UI output is truncated when I use a workflow that incorporates Tensorflow output.

Could be the particular formatting of the training that stops the output to the UI? The logs are properly shown in the Argo UI exactly to the point before Tensorflow training output starts, (example training line from minio logs below):
`1501/1501 [==============================] - 57s 33ms/step - loss: 0.8667 - accuracy: 0.7131 - precision: 0.8383 - recall: 0.6788 - f1_score: 0.7311 - val_loss: 0.1472 - val_accuracy: 0.9567 - val_precision: 1.0000 - val_recall: 0.9786 - val_f1_score: 0.9890`

## Diagnostics

What Kubernetes provider are you using? 
Kubeadm 1.19.7

What version of Argo Workflows are you running? 
v2.12.8

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by simster7** [comment]
Might very well be that our current loggin infrastructure doesn't deal well with `\r`, which is what is used to replace the console output line to get that nice "progress effect" that tensorflow displays in lines like those.

I'll look into it

**Comment by simster7** [comment]
In fact, thinking this a bit more I think this is likely the case

**Comment by simster7** [comment]
Will have to see if this is something we can deal with or if K8s' Pod log streaming interface is the problem

**Comment by alexec** [comment]
Bug missing reproduction step. Can you please attach a workflow that reproduces the problem that we can run?

**Comment by simster7** [comment]
My suspicions were right: the issue is the use of `\r` _and_ K8s itself does not handle it. Since we only display logs that are passed to use by K8s, I'm not sure what we can do about this.

See below for reproduction steps:

Workflow
```yaml
# script templates provide a way to run arbitary snippets of code
# in any language, to produce a output "result" via the standard out
# of the template. Results can then be referenced using the variable,
# {{steps.<stepname>.outputs.result}}, and used as parameter to other
# templates, and in 'when', and 'withParam' clauses.
# This example demonstrates the use of a python script to
# generate a random number which is printed in the next step.
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: scripts-python-
spec:
  entrypoint: python-script-example
  templates:
  - name: python-script-example
    steps:
    - - name: generate
        template: gen-random-int

  - name: gen-random-int
    script:
      image: python:alpine3.6
      command: [python]
      source: |
        import time
        print("Starting job")
        for i in range(10):
          print("Progress", i, "/ 10", end="\r")
          time.sleep(1)
        print("Ended job")
```

K8s doesn't display any logs while pod is still running:

![image](https://user-images.githubusercontent.com/614838/108535126-d38c2900-728f-11eb-9299-d6b27f09bfaf.png)


Final K8s Logs:

![image](https://user-images.githubusercontent.com/614838/108534881-7f814480-728f-11eb-85fa-9712f11c0c47.png)


Final UI Logs:
![image](https://user-images.githubusercontent.com/614838/108534799-69738400-728f-11eb-800b-bd8a4e847efd.png)

**Comment by simster7** [comment]
Hey @Atharex I actually got something for you to try. When running `python`, try the `-u` flag: `python -u`. Example from another community member: https://github.com/onepanelio/templates/blob/d23d2e1a2e49dce37a4e9a428e3205b930a190c2/workflows/tensorflow-mnist-training/template.yaml#L29

Let me know if this works

**Comment by Atharex** [comment]
Thanks for the hint, seems to work now! Turned out the problem was also a bit trickier. The node I was running the workload on was in a different subnet than other nodes. After it was transferred to the same subnet, the logs started showing up normally.

**Comment by alexec** [comment]
Closing as invalid. Please comment if incorrect.

**Comment by Erotemic** [comment]
As a "nice to have" it might be good if argo logs gave a warning that this behavior existed. I find it very non-intuitive. When I ask for logs, I expect to get all of stdout, so even though it is a K8s problem, if there is a workaround, or at least a warning that points to workarounds, that might make the experience of using argo nicer.
